crates/sui-types/src/effects/mod.rs

```diff
@@ -126,8 +132,79 @@ impl TransactionEffects {
         events_digest: Option<TransactionEventsDigest>,
         dependencies: Vec<TransactionDigest>,
     ) -> Self {
-        // TODO: when there are multiple versions, use protocol_version to construct the
-        // appropriate one.
+        Self::V1(TransactionEffectsV1::new(
+            status,
+            executed_epoch,
+            gas_used,
+            modified_at_versions,
+            shared_objects,
+            transaction_digest,
+            created,
+            mutated,
+            unwrapped,
+            deleted,
+            unwrapped_then_deleted,
+            wrapped,
+            gas_object,
+            events_digest,
+            dependencies,
+        ))
+    }
+
+    /// Creates a TransactionEffects message from the results of execution, choosing the correct
+    /// format for the current protocol version.
+    pub fn new_from_execution_v2(
+        _protocol_version: ProtocolVersion,
+        status: ExecutionStatus,
+        executed_epoch: EpochId,
+        gas_used: GasCostSummary,
+        shared_objects: Vec<ObjectRef>,
+        transaction_digest: TransactionDigest,
+        lamport_version: SequenceNumber,
+        object_changes: BTreeMap<ObjectID, EffectsObjectChange>,
+        gas_object: (ObjectRef, Owner),
+        events_digest: Option<TransactionEventsDigest>,
+        dependencies: Vec<TransactionDigest>,
+    ) -> Self {
+        let mut created = vec![];
+        let mut mutated = vec![];
+        let mut unwrapped = vec![];
+        let mut deleted = vec![];
+        let mut unwrapped_then_deleted = vec![];
+        let mut wrapped = vec![];
+        let mut modified_at_versions = vec![];
+        let mut deleted_at_versions = vec![];
+        for (id, change) in object_changes {
+            if let Some(((version, digest), owner)) = change.get_written_object(lamport_version) {
```

Comment:
Would it be better to match on the `change` directly? It feels a bit strange to if/else on the `is_<case>` and then assert instead of just matching

Reply:
Good point. Let me fix this.